File: packages/typescript/test/native-sidebar.test.ts
import { describe, expect, it } from 'bun:test'
import { CraftApp, createApp, type AppConfig } from '../src/index'

describe('Native Sidebar', () => {
  describe('createApp with native sidebar config', () => {
    it('should accept nativeSidebar option', () => {
      const app = createApp({
        url: 'http://localhost:3456/app',
        window: {
          nativeSidebar: true,
          sidebarWidth: 240,
        },
      })
      expect(app).toBeInstanceOf(CraftApp)
    })

    it('should accept sidebarConfig with sections', () => {
      const sidebarConfig = {
        sections: [
          {
            id: 'home',
            title: 'Home',
            items: [
              { id: 'dashboard', label: 'Dashboard', icon: 'house.fill', url: '/dashboard' },
            ],
          },
        ],
        minWidth: 200,
        maxWidth: 280,
      }

      const app = createApp({
        url: 'http://localhost:3456/app',
        window: {
          nativeSidebar: true,
          sidebarWidth: 240,
          sidebarConfig,
        },
      })
      expect(app).toBeInstanceOf(CraftApp)
    })

    it('should work without sidebarConfig', () => {
      const app = createApp({
        url: 'http://localhost:3456/app',
        window: {
          nativeSidebar: true,
        },
      })
      expect(app).toBeInstanceOf(CraftApp)
    })

    it('should default nativeSidebar to false', () => {
      const app = createApp({
        url: 'http://localhost:3456/app',
      })
      expect(app).toBeInstanceOf(CraftApp)
    })
  })

  describe('native sidebar with dashboard-like config', () => {
    it('should handle full dashboard configuration', () => {
      const config: AppConfig = {
        url: 'http://localhost:3456/app?native-sidebar=1',
        window: {
          title: 'Stacks Dashboard',
          width: 1400,
          height: 900,
          titlebarHidden: true,
          nativeSidebar: true,
          sidebarWidth: 240,
          sidebarConfig: {
            sections: [
              {
                id: 'home',
                title: 'Home',
                items: [
                  { id: 'home', label: 'Dashboard', icon: 'house.fill', url: '/pages/index' },
                ],
              },
              {
                id: 'data',
                title: 'Data',
                items: [
                  { id: 'data-dashboard', label: 'Dashboard', icon: 'gauge', url: '/pages/data/dashboard' },
                  { id: 'model-user', label: 'User', icon: 'person.fill', url: '/pages/data/user' },
                ],
              },
            ],
            minWidth: 200,
            maxWidth: 280,
          },
        },
      }

      const app = createApp(config)
      expect(app).toBeInstanceOf(CraftApp)
    })

    it('should handle URL with native-sidebar query param', () => {
      const app = createApp({
        url: 'http://localhost:3456/app?native-sidebar=1',
        window: {
          nativeSidebar: true,
        },
      })
      expect(app).toBeInstanceOf(CraftApp)
    })
  })

  describe('sidebar config validation', () => {
    it('should handle empty sections array', () => {
      const app = createApp({
        url: 'http://localhost:3456/app',
        window: {
          nativeSidebar: true,
          sidebarConfig: {
            sections: [],
          },
        },
      })
      expect(app).toBeInstanceOf(CraftApp)
    })

    it('should handle sections with multiple items', () => {
      const app = createApp({
        url: 'http://localhost:3456/app',
        window: {
          nativeSidebar: true,
          sidebarConfig: {
            sections: [
              {
                id: 'content',
                title: 'Content',
                items: [
                  { id: 'pages', label: 'Pages', icon: 'doc.fill', url: '/pages' },
                  { id: 'posts', label: 'Posts', icon: 'text.bubble.fill', url: '/posts' },
                  { id: 'tags', label: 'Tags', icon: 'tag', url: '/tags' },
                  { id: 'categories', label: 'Categories', icon: 'tag.fill', url: '/categories' },
                ],
              },
            ],
          },
        },
      })
      expect(app).toBeInstanceOf(CraftApp)
    })

    it('should handle custom sidebar widths', () => {
      const narrow = createApp({
        url: 'http://localhost:3456/app',
        window: { nativeSidebar: true, sidebarWidth: 150 },
      })
      const wide = createApp({
        url: 'http://localhost:3456/app',
        window: { nativeSidebar: true, sidebarWidth: 400 },
      })
      expect(narrow).toBeInstanceOf(CraftApp)
      expect(wide).toBeInstanceOf(CraftApp)
    })
  })

  describe('close with native sidebar', () => {
    it('should not throw when closing native sidebar app', () => {
      const app = createApp({
        url: 'http://localhost:3456/app',
        window: {
          nativeSidebar: true,
          sidebarWidth: 240,
          sidebarConfig: {
            sections: [
              {
                id: 'test',
                title: 'Test',
                items: [{ id: 'item', label: 'Item', icon: 'star', url: '/test' }],
              },
            ],
          },
        },
      })
      expect(() => app.close()).not.toThrow()
    })
  })
})

File: packages/zig/test/native_sidebar_test.zig
const std = @import("std");
const testing = std.testing;
const cli = @import("../src/cli.zig");

// ============================================
// Native Sidebar CLI Options
// ============================================

test "WindowOptions - native_sidebar defaults to false" {
    const options = cli.WindowOptions{};
    try testing.expect(!options.native_sidebar);
}

test "WindowOptions - native_sidebar can be enabled" {
    const options = cli.WindowOptions{
        .native_sidebar = true,
    };
    try testing.expect(options.native_sidebar);
}

test "WindowOptions - sidebar_width defaults to 220" {
    const options = cli.WindowOptions{};
    try testing.expectEqual(@as(u32, 220), options.sidebar_width);
}

test "WindowOptions - sidebar_width can be customized" {
    const options = cli.WindowOptions{
        .sidebar_width = 240,
    };
    try testing.expectEqual(@as(u32, 240), options.sidebar_width);
}

test "WindowOptions - sidebar_width narrow" {
    const options = cli.WindowOptions{
        .sidebar_width = 150,
    };
    try testing.expectEqual(@as(u32, 150), options.sidebar_width);
}

test "WindowOptions - sidebar_width wide" {
    const options = cli.WindowOptions{
        .sidebar_width = 400,
    };
    try testing.expectEqual(@as(u32, 400), options.sidebar_width);
}

test "WindowOptions - sidebar_config defaults to null" {
    const options = cli.WindowOptions{};
    try testing.expect(options.sidebar_config == null);
}

test "WindowOptions - sidebar_config can be set" {
    const config = "{\"sections\":[]}";
    const options = cli.WindowOptions{
        .sidebar_config = config,
    };
    try testing.expect(options.sidebar_config != null);
    try testing.expectEqualStrings("{\"sections\":[]}", options.sidebar_config.?);
}

// ============================================
// Native Sidebar with URL mode
// ============================================

test "WindowOptions - native sidebar with URL" {
    const options = cli.WindowOptions{
        .url = "http://localhost:3456/app?native-sidebar=1",
        .native_sidebar = true,
        .sidebar_width = 240,
    };

    try testing.expect(options.native_sidebar);
    try testing.expectEqual(@as(u32, 240), options.sidebar_width);
    try testing.expectEqualStrings("http://localhost:3456/app?native-sidebar=1", options.url.?);
}

test "WindowOptions - native sidebar with HTML" {
    const options = cli.WindowOptions{
        .html = "<h1>Dashboard</h1>",
        .native_sidebar = true,
        .sidebar_width = 260,
    };

    try testing.expect(options.native_sidebar);
    try testing.expectEqual(@as(u32, 260), options.sidebar_width);
    try testing.expectEqualStrings("<h1>Dashboard</h1>", options.html.?);
}

// ============================================
// Full dashboard-like configuration
// ============================================

test "WindowOptions - full dashboard config" {
    const sidebar_json =
        \\{"sections":[{"id":"home","title":"Home","items":[{"id":"home","label":"Dashboard","icon":"house.fill","url":"/pages/index"}]}]}
    ;

    const options = cli.WindowOptions{
        .url = "http://localhost:3456/app?native-sidebar=1",
        .title = "Stacks Dashboard",
        .width = 1400,
        .height = 900,
        .titlebar_hidden = true,
        .native_sidebar = true,
        .sidebar_width = 240,
        .sidebar_config = sidebar_json,
    };

    try testing.expectEqualStrings("Stacks Dashboard", options.title);
    try testing.expectEqual(@as(u32, 1400), options.width);
    try testing.expectEqual(@as(u32, 900), options.height);
    try testing.expect(options.titlebar_hidden);
    try testing.expect(options.native_sidebar);
    try testing.expectEqual(@as(u32, 240), options.sidebar_width);
    try testing.expect(options.sidebar_config != null);
}

test "WindowOptions - native sidebar does not imply system tray" {
    const options = cli.WindowOptions{
        .native_sidebar = true,
    };

    try testing.expect(options.native_sidebar);
    try testing.expect(!options.system_tray);
    try testing.expect(!options.hide_dock_icon);
}

test "WindowOptions - native sidebar with dark mode" {
    const options = cli.WindowOptions{
        .native_sidebar = true,
        .dark_mode = true,
    };

    try testing.expect(options.native_sidebar);
    try testing.expectEqual(true, options.dark_mode.?);
}

test "WindowOptions - native sidebar with light mode" {
    const options = cli.WindowOptions{
        .native_sidebar = true,
        .dark_mode = false,
    };

    try testing.expect(options.native_sidebar);
    try testing.expectEqual(false, options.dark_mode.?);
}

// ============================================
// CLI argument parsing
// ============================================

test "parseArgs - native-sidebar flag" {
    const allocator = testing.allocator;
    const args = [_][:0]const u8{ "craft", "--native-sidebar", "--url", "http://localhost:3456" };
    const options = try cli.parseArgs(allocator, &args);

    try testing.expect(options.native_sidebar);
    try testing.expectEqualStrings("http://localhost:3456", options.url.?);

    allocator.free(options.url.?);
}

test "parseArgs - sidebar-width flag" {
    const allocator = testing.allocator;
    const args = [_][:0]const u8{ "craft", "--native-sidebar", "--sidebar-width", "240", "--url", "http://localhost:3456" };
    const options = try cli.parseArgs(allocator, &args);

    try testing.expect(options.native_sidebar);
    try testing.expectEqual(@as(u32, 240), options.sidebar_width);

    allocator.free(options.url.?);
}

test "parseArgs - sidebar-config flag" {
    const allocator = testing.allocator;
    const config_json = "{\"sections\":[]}";
    const args = [_][:0]const u8{ "craft", "--native-sidebar", "--sidebar-config", config_json, "--url", "http://localhost:3456" };
    const options = try cli.parseArgs(allocator, &args);

    try testing.expect(options.native_sidebar);
    try testing.expect(options.sidebar_config != null);
    try testing.expectEqualStrings("{\"sections\":[]}", options.sidebar_config.?);

    allocator.free(options.url.?);
    allocator.free(options.sidebar_config.?);
}

test "parseArgs - all native sidebar flags together" {
    const allocator = testing.allocator;
    const config_json = "{\"sections\":[{\"id\":\"home\"}]}";
    const args = [_][:0]const u8{
        "craft",
        "--url",
        "http://localhost:3456/app?native-sidebar=1",
        "--title",
        "Dashboard",
        "--width",
        "1400",
        "--height",
        "900",
        "--titlebar-hidden",
        "--native-sidebar",
        "--sidebar-width",
        "240",
        "--sidebar-config",
        config_json,
    };
    const options = try cli.parseArgs(allocator, &args);

    try testing.expectEqualStrings("http://localhost:3456/app?native-sidebar=1", options.url.?);
    try testing.expectEqualStrings("Dashboard", options.title);
    try testing.expectEqual(@as(u32, 1400), options.width);
    try testing.expectEqual(@as(u32, 900), options.height);
    try testing.expect(options.titlebar_hidden);
    try testing.expect(options.native_sidebar);
    try testing.expectEqual(@as(u32, 240), options.sidebar_width);
    try testing.expect(options.sidebar_config != null);

    allocator.free(options.url.?);
    allocator.free(options.title);
    allocator.free(options.sidebar_config.?);
}

test "parseArgs - sidebar-width without native-sidebar" {
    const allocator = testing.allocator;
    const args = [_][:0]const u8{ "craft", "--sidebar-width", "300", "--url", "http://localhost:3456" };
    const options = try cli.parseArgs(allocator, &args);

    // sidebar-width is set but native_sidebar is still false
    try testing.expect(!options.native_sidebar);
    try testing.expectEqual(@as(u32, 300), options.sidebar_width);

    allocator.free(options.url.?);
}

test "parseArgs - default sidebar-width when native-sidebar enabled" {
    const allocator = testing.allocator;
    const args = [_][:0]const u8{ "craft", "--native-sidebar", "--url", "http://localhost:3456" };
    const options = try cli.parseArgs(allocator, &args);

    try testing.expect(options.native_sidebar);
    try testing.expectEqual(@as(u32, 220), options.sidebar_width);

    allocator.free(options.url.?);
}
